File: src/내리막길.js
const input = require('fs').readFileSync('./example.txt').toString().split('\n');
const [N, M] = input.shift().split(' ').map(v => +v);
const arr = input.map(row => row.split(" ").map(v => +v));



function solution(N, M, arr) {
  const dx = [1, 0, -1, 0];
  const dy = [0, 1, 0, -1];
  const visited = Array.from(Array(N), () => Array(M).fill(false));

  const rangeCheck = (row, col) => (row >= 0 && col >= 0 && row < N && col < M);

  const dfs = (row, col) => {
    if (row === N - 1 && col === M - 1) return 1;

    if (!visited[row][col]) {
      visited[row][col] = 0;
      for (let i = 0; i < 4; i++) {
        const [nextRow, nextCol] = [row + dy[i], col + dx[i]];
        if (rangeCheck(nextRow, nextCol) && (arr[nextRow][nextCol] < arr[row][col])) {
          visited[row][col] += dfs(nextRow, nextCol);
        }
      }
    }
    return visited[row][col];
  }

  dfs(0, 0);
  console.log(visited[0][0]);
}

solution(N, M, arr);

File: src/내리막길.cpp
#include <bits/stdc++.h>

using namespace std;

int arr[500][500];
int visited[500][500];
int M, N;
const int dx[4] = {1, 0, -1, 0};
const int dy[4] = {0, 1, 0, -1};

bool rangeCheck(int row, int col)
{
  return (row >= 0 && col >= 0 && row < M && col < N);
};

int dfs(int row, int col)
{
  if (row == M - 1 && col == N - 1)
    return 1;

  if (visited[row][col] == -1)
  {
    visited[row][col] = 0;
    for (int i = 0; i < 4; i++)
    {
      int nextRow = row + dy[i];
      int nextCol = col + dx[i];

      if (rangeCheck(nextRow, nextCol) && (arr[nextRow][nextCol] < arr[row][col]))
      {
        visited[row][col] += dfs(nextRow, nextCol);
      }
    }
  }
  return visited[row][col];
}

int main()
{
  cin >> M >> N;
  for (int i = 0; i < M; i++)
  {
    for (int j = 0; j < N; j++)
    {
      cin >> arr[i][j];
      visited[i][j] = -1;
    }
  }

  dfs(0, 0);
  cout << visited[0][0];

  return 0;
}